### PR TITLE
Make GitPackCache include ObjectType

### PR DESCRIPTION
**Pull request by georg-jung**
See https://github.com/dotnet/Nerdbank.GitVersioning/pull/876#discussion_r1146110974

This adds caching of the determined objectType alongside the object's content. While the approach in d0aa014 seems a bit more straight forward than the one in fafa498, I think it is preferrable to have the contentType as part of the cached value instead of the cache key. Otherwise, repeated requests with a mismatching objectType will not profit from caching.

I thought about adding proper `TryGetObject` implementations as part of this. This turned out to be quite hard to get right though. The topics aren't too closely related too so I think they work as seperate PRs. Because these cache improvements should be sufficient to make #876 work, I decided to split this up and postpone the proper `TryGetObject` implementations.

### CODE REVIEW

File: src/NerdBank.GitVersioning/ManagedGit/GitPackCache.cs [hunk 1]
```diff
@@ -45,14 +45,15 @@ public abstract class GitPackCache : IDisposable
...
+    /// <param name="objectType">The object type of the object to add to the cache.</param>
     /// <param name="stream">
     /// A <see cref="Stream"/> which represents the object to add. This stream
     /// will be copied to the cache.
     /// </param>
     /// <returns>
     /// A <see cref="Stream"/> which represents the cached entry.
     /// </returns>
-    public abstract Stream Add(long offset, Stream stream);
+    public abstract Stream Add(long offset, string objectType, Stream stream);
```
Comment (ronaldbarendse) [inline src/NerdBank.GitVersioning/ManagedGit/GitPackCache.cs:56]: I'd suggest aligning the parameter order with the tuple: `Add(long offset, Stream stream, string objectType)`.

**Comment by georg-jung** [comment]
@AArnott any thoughts on this? I got some time in the coming weeks to work on this or #876 if there's something to improve (if this was merged, #876 would need to be rebased obviously).

**Review by AArnott — APPROVED** [review]
Looks fantastic! Thank you.

**Comment by AArnott** [comment]
Sorry it took so long for me to get to this, @georg-jung.

**Comment by AArnott** [comment]
So weird.... I've re-queued the build 3 times and the test process crashes every time. Yet several other builds today passed the first time, so there seems to be something uniquely about this PR that causes a test crash, but it sure seems safe enough. I'm still looking into it.

**Comment by georg-jung** [comment]
Thanks for taking a look!

Interestingly, when I re-run the pipeline on the branch in my fork it doesn't fail: <https://dev.azure.com/georg-jung/Nerdbank.GitVersioning/_build/results?buildId=1870&view=results>

(Well the pipeline *did* fail, but that's because it can't publish packages to your CI feed; the tests ran successfully)

**Comment by georg-jung** [comment]
Thanks for taking a look!

Back then, my pipeline ran. I started a run on my ADO on this PR's branch here: <https://dev.azure.com/georg-jung/Nerdbank.GitVersioning/_build/results?buildId=1871&view=results>

I also rebased the PR's contents in a second branch to be on-par with latest main and starte a run on it too: <https://dev.azure.com/georg-jung/Nerdbank.GitVersioning/_build/results?buildId=1872&view=results>

Both are still in progress as of writing.

(Well the pipeline *did*/*does* fail, but that's because it can't publish packages to your CI feed; the tests ran successfully)

**Comment by georg-jung** [comment]
All the test steps seem to work in my ADO, for the rebased version as well as the original one. Really strange.

**Pull request merged**